### DISCUSSION TITLE
AttributeError: 'async_generator' object has no attribute 'config_entries'

### DISCUSSION THREAD

**Discussion by masaccio** [MatthewFlamm/pytest-homeassistant-custom-component#160] [Q&A] [answered]
I was testing against HA core 2022.12.4 and pytest-homeassistant-custom-component 0.12.33 and my custom integration broke this week with the latest HA core update.

I am in the process of fixing and have got past Poetry dependency problems but pytest-homeassistant-custom-component is failing in odd ways:

``` python
hass = <async_generator object _hass at 0x11b9f35b0>, bypass_get_data = None

    @pytest.mark.asyncio
    async def test_successful_config_flow(hass, bypass_get_data):
        """Test a successful config flow."""
        # Initialize a config flow
>       result = await hass.config_entries.flow.async_init(
            DOMAIN, context={"source": config_entries.SOURCE_USER}
        )
E       AttributeError: 'async_generator' object has no attribute 'config_entries'

tests/test_config_flow.py:35: AttributeError
```

This feels like I am doing something bad so before I raise an issue with better detail, I thought I would ask here. Possibly related, this existing fixture in `contest.py`:

``` python
@pytest.fixture(autouse=True)
def auto_enable_custom_integrations(enable_custom_integrations):
    yield
```

also generates similar errors about the `hass` object:

``` python
hass = <async_generator object _hass at 0x11f527cd0>

    @pytest.fixture
    def enable_custom_integrations(hass: HomeAssistant) -> None:
        """Enable custom integrations defined in the test dir."""
>       hass.data.pop(loader.DATA_CUSTOM_COMPONENTS)
E       AttributeError: 'async_generator' object has no attribute 'data'
```

**Comment by MatthewFlamm** [comment]
This is the second report of this problem. See #158.  There it was found that it is testing dependent, i.e. passes in some environments but not others.  Do you observe this as well?

I'm not sure if this is a problem in the code that is now being uncovered, or maybe it is an order of fixture execution that might have changed.  The tests in this package are not affected by this problem, which might be a starting point for investigating differences.

**Reply by MatthewFlamm** [reply]
cc @Shulyaka

**Reply by Shulyaka** [reply]
My best guess is that it is some kind of issue with `pytest` itself, its logic of auto guessing sync/async fixtures breaks if all tests are async. Like if all the tests are async it incorrectly assumes all fixtures are also async.
My error gone away after I added one sync test function.

**Reply by masaccio** [reply, marked as answer]
Thanks both - adding `asyncio_mode = "auto"` to my `pyproject.toml` was what I needed. It also lets me remove the various `@pytest.mark.asyncio` lines. Reading through some Stackoverflow threads, fixtures with pytest and asyncio does need to be handled with care, but that `asyncio_mode` is the cleanest answer.

**Reply by MatthewFlamm** [reply]
Thanks for following up on the fix.

**Reply by sdwilsh** [reply]
For some additional clarity, the specific bits you add to `pyproject.toml` should look like these:
```
[tool.pytest.ini_options]
asyncio_mode = "auto"
```
